plugins/amqp_compressed_proof_plugin/amqp_compressed_proof_plugin.cpp

```diff
@@ -0,0 +1,397 @@
+          act.authorization.size() != 1)
+         return false;
+      auto& auth = act.authorization[0];
+      return auth.actor == eosio::chain::config::system_account_name &&
+             auth.permission == eosio::chain::config::active_name;
+   }
+
+   std::map<transaction_id_type, chain::transaction_trace_ptr> cached_traces;
+   std::optional<chain::transaction_trace_ptr>                 onblock_trace;
+   std::vector<std::pair<compressed_proof_generator::action_filter_func,
+                         compressed_proof_generator::merkle_proof_result_func>>  callbacks;
+
+   fc::optional<boost::filesystem::path> reversible_path;
+};
+
+using generator_impl = compressed_proof_generator_impl;
+
+compressed_proof_generator::compressed_proof_generator() : my(new compressed_proof_generator_impl) {}
+
+compressed_proof_generator::compressed_proof_generator(const boost::filesystem::path& p) : my(new compressed_proof_generator_impl) {
+   my->reversible_path = p;
+
+   if(boost::filesystem::exists(*my->reversible_path)) {
+      try {
+         fc::datastream<fc::cfile> file;
+         file.set_file_path(*my->reversible_path);
+         file.open("rb");
+         fc::unsigned_int count;
+         fc::raw::unpack(file, count);
+         while(count.value--) {
+            compressed_proof_generator_impl::reversible_action_entries entry;
+            fc::raw::unpack(file, entry);
+            my->reversible_action_entries_index.emplace(entry);
+         }
+      } FC_RETHROW_EXCEPTIONS(error, "Failed to load compressed merkle generator reversible state file ${f}", ("f", (fc::path)*my->reversible_path));
+   }
+   else {
+      try {
+         boost::filesystem::ofstream s(*my->reversible_path);
+      } FC_RETHROW_EXCEPTIONS(error, "Unable to create compressed merkle generator reversible state file ${f}", ("f", (fc::path)*my->reversible_path));
+   }
+   boost::system::error_code ec;
+   boost::filesystem::remove(*my->reversible_path, ec);
+}
+
+compressed_proof_generator::~compressed_proof_generator() {
+   if(my->reversible_path) {
+      try {
+         fc::unsigned_int num = my->reversible_action_entries_index.size();
+         fc::datastream<fc::cfile> file;
+         file.set_file_path(*my->reversible_path);
+         file.open("wb");
+         fc::raw::pack(file, num);
+         for(const auto& v : my->reversible_action_entries_index)
+            fc::raw::pack(file, v);
+      } FC_LOG_AND_DROP();
+   }
+}
+
+void compressed_proof_generator::on_applied_transaction(const chain::transaction_trace_ptr& trace) {
+   if(!trace->receipt)
+      return;
+   //only executed & delayed transaction traces would make it in to action_mroot
+   if(trace->receipt->status != chain::transaction_receipt::status_enum::executed &&
+      trace->receipt->status != chain::transaction_receipt::status_enum::delayed)
+      return;
+
+   if(my->is_onblock(trace))
+      my->onblock_trace.emplace(trace);
```

Comment:
What if we abort a block in which onblock suceeded and start a new block in which onblock failed?

Reply:
I believe state history & trace api have the same semantics? Honestly I didn't give it much thought since I figured those had been noodled through.

Reply:
I think this is a problem here and in state history and trace as well, but only when running the chain in speculative mode (which we probably don't want to do anyway?).  I suppose it could also be a problem if we receive a bad block in head mode.  If we get a bad block in irreversible mode, we have bigger problems.

Reply:
@spoonincode I will create a fix for this here, state history, and trace.

Reply:
Fix pushed up to this branch.

Reply:
This is turning out to be more complicated than I first thought, still looking into it. `block_num` is available but could be the same across an aborted block. id is not available unless block is from an external producer. There is no start_block or abort_block signal.

Reply:
Added verification of block_num see discussion below.